File: leetcode/8-string-to-integer-atoi/main.go
package main

import (
	"fmt"
	"math"
	"strings"
)

/*
	1st approach
	- straight forward approach
	- be careful of the corner cases
	Time 	O(n) n=length of input string
	Space O(1)
	4ms beats 100%
	15jan2019
*/
func myAtoi0(str string) int {
	sum := 0
	negative := 1
	str = strings.TrimSpace(str)
	for i := 0; i < len(str); i++ {
		if i == 0 {
			if str[i] == '+' {
				continue
			} else if str[i] == '-' {
				negative = -1
			} else if str[i]-'0' <= 9 {
				sum = int(str[i] - '0')
			} else {
				break
			}
		} else {
			if str[i]-'0' <= 9 {
				sum = sum*10 + int(str[i]-'0')
				temp := sum * negative
				if temp < math.MinInt32 {
					return math.MinInt32
				}
				if temp > math.MaxInt32 {
					return math.MaxInt32
				}
			} else {
				break
			}
		}
	}
	result := sum * negative
	if result < math.MinInt32 {
		return math.MinInt32
	}
	if result > math.MaxInt32 {
		return math.MaxInt32
	}
	return result
}

/*
	2nd approach is actaully a rewritten form of 1st
	Time 	O(n) n=length of input string
	Space O(1)
	4ms beats 100%
	15jan2019
*/
func myAtoi(str string) int {
	sum := 0
	sign := 1
	str = strings.TrimSpace(str)
	for i := 0; i < len(str); i++ {
		if i == 0 {
			if str[i] == '+' {
				continue
			} else if str[i] == '-' {
				sign = -1
			} else if str[i]-'0' <= 9 {
				sum = int(str[i] - '0')
			} else {
				break
			}
		} else {
			if str[i]-'0' <= 9 {
				sum = sum*10 + int(str[i]-'0')
				c := checkBoundries(sum, sign)
				abs := c
				if c < 0 {
					abs = c * -1
				}
				if abs != sum {
					return c
				}
			} else {
				break
			}
		}
	}
	return checkBoundries(sum, sign)
}

func checkBoundries(num int, sign int) int {
	result := num * sign
	if result < math.MinInt32 {
		return math.MinInt32
	}
	if result > math.MaxInt32 {
		return math.MaxInt32
	}
	return result
}

func main() {
	fmt.Println(myAtoi("42"))
	fmt.Println(myAtoi("   -42"))
	fmt.Println(myAtoi("4193 with words"))
	fmt.Println(myAtoi("words and 987"))
	fmt.Println(myAtoi("22 and 987"))
	fmt.Println(myAtoi(" 22 and 987"))
	fmt.Println(myAtoi(" -22 and 987"))
	fmt.Println(myAtoi(" -+22 and 987"))
	fmt.Println(myAtoi("-91283472332"))
	fmt.Println(myAtoi(" -91283472332"))
	fmt.Println(myAtoi("91283472332"))
	fmt.Println(myAtoi(" +91283472332"))
	fmt.Println(myAtoi(" -91283472332 a"))
	fmt.Println(myAtoi(" a-91283472332"))
	fmt.Println(myAtoi("9223372036854775808"))
}

File: leetcode/8-string-to-integer-atoi/main.py
"""
    1st approach
	- straight forward approach
	- be careful of the corner cases

	Time 	O(n) n=length of input string
	Space O(1)
	4ms beats 100%
	19may2019
"""


class Solution(object):
    def myAtoi(self, s):
        """
        :type s: str
        :rtype: int
        """
        s = s.strip()
        if len(s) == 0:
            return 0
        num = 0
        polarity = 1
        if s[0] == '+':
            pass
        elif s[0] == '-':
            polarity = -1
        elif s[0].isdigit():
            num = int(s[0])
        else:
            return 0

        for i in range(1, len(s)):
            c = s[i]
            if c.isdigit():
                num = num*10 + int(c)
            else:
                break
        return self.bound(num * polarity)

    def bound(self, num):
        if num < -2**31:
            return -2**31
        elif num > 2**31-1:
            return 2**31-1
        return num


s = Solution()

print(s.myAtoi('42'))
print(s.myAtoi('-42'))
print(s.myAtoi(' -42'))
print(s.myAtoi('4193 with words'))
print(s.myAtoi('words and 987'))
print(s.myAtoi('22 and 987'))
print(s.myAtoi(' -22 and 987'))
print(s.myAtoi(' -+22 and 987'))
print(s.myAtoi('-91283472332'))
print(s.myAtoi('-91283472332 a'))
print(s.myAtoi(' -91283472332'))
print(s.myAtoi(' -91283472332 a'))
print(s.myAtoi(' a-91283472332'))
print(s.myAtoi('00000-42a1234'))

print("----")


class Solution:
    def myAtoi(self, s: str) -> int:
        if len(s) == 0:
            return 0
        numStr = ''
        isPossitveSet = False
        isNegative = False
        for i in range(len(s)):
            c = s[i]
            if c == '-':
                if isNegative or isPossitveSet or len(numStr) > 0:
                    break
                isNegative = True
            elif c == '+':
                if isNegative or isPossitveSet or len(numStr) > 0:
                    break
                isPossitveSet = True
            elif c.isdigit():
                numStr += c
            elif c == ' ':
                if isNegative or isPossitveSet or len(numStr) > 0:
                    break
            else:
                break
        if len(numStr) == 0:
            return 0
        return self.bound(int(numStr), isNegative)

    def bound(self, x, isNegative):
        if isNegative:
            x = -x
        x = max(x, -2**31)
        x = min(2**31-1, x)
        return x


s = Solution()

print(s.myAtoi('42'))
print(s.myAtoi('-42'))
print(s.myAtoi(' -42'))
print(s.myAtoi('4193 with words'))
print(s.myAtoi('words and 987'))
print(s.myAtoi('22 and 987'))
print(s.myAtoi(' -22 and 987'))
print(s.myAtoi(' -+22 and 987'))
print(s.myAtoi('-91283472332'))
print(s.myAtoi('-91283472332 a'))
print(s.myAtoi(' -91283472332'))
print(s.myAtoi(' -91283472332 a'))
print(s.myAtoi(' a-91283472332'))
print(s.myAtoi('00000-42a1234'))
